core/jvm/src/main/scala/fs2/interop/flow/package.scala

```diff
@@ -0,0 +1,253 @@
+    *                    the actual `subscribe` operation; as another function receiving the [[Publisher]].
+    *                    The `subscribe` operation must be called exactly once.
+    */
+  def fromOnSubscribe[F[_], A](
+      bufferSize: Int
+  )(
+      onSubscribe: (Publisher[A] => F[Unit]) => F[Unit]
+  )(implicit
+      F: Async[F]
+  ): Stream[F, A] =
+    fromSubscriber[F, A](bufferSize) { subscriber =>
+      val subscribe: Publisher[A] => F[Unit] =
+        publisher => F.delay(publisher.subscribe(subscriber))
+
+      onSubscribe(subscribe)
+    }
+
+  /** Creates a [[Stream]] from an [[Publisher]].
+    *
+    * @note The publisher only receives a subscriber when the stream is run.
+    *
+    * @param publisher The [[Publisher]] to transform.
+    * @param bufferSize setup the number of elements asked each time from the [[Publisher]].
+    *                   A high number can be useful if the publisher is triggering from IO,
+    *                   like requesting elements from a database.
+    *                   The publisher can use this `bufferSize` to query elements in batch.
+    *                   A high number will also lead to more elements in memory.
+    */
+  def fromPublisher[F[_], A](
+      publisher: Publisher[A],
+      bufferSize: Int
+  )(implicit
+      F: Async[F]
+  ): Stream[F, A] =
+    fromOnSubscribe[F, A](bufferSize) { subscribe =>
+      subscribe(publisher)
+    }
+
+  implicit final class PublisherOps[A](private val publisher: Publisher[A]) extends AnyVal {
+
+    /** Creates a [[Stream]] from an [[Publisher]].
+      *
+      * @param bufferSize setup the number of elements asked each time from the [[Publisher]].
+      *                   A high number can be useful is the publisher is triggering from IO,
+      *                   like requesting elements from a database.
+      *                   The publisher can use this `bufferSize` to query elements in batch.
+      *                   A high number will also lead to more elements in memory.
+      */
+    def toStream[F[_]](bufferSize: Int)(implicit F: Async[F]): Stream[F, A] =
+      flow.fromPublisher(publisher, bufferSize)
+  }
+
+  /** Creates a [[Publisher]] from a [[Stream]].
+    *
+    * This publisher can only have a single subscription.
+    * The stream is only ran when elements are requested.
+    * Closing the [[Resource]] means gracefully shutting down the subscription,
+    * Thus, no more elements will be published.
+    *
+    * @see [[subscribeStream]] for a simpler version that only requires a [[Subscriber]].
+    *
+    * @param stream The [[Stream]] to transform.
+    * @return A tuple of the [[Publisher]]
+    *         and a [[DeferredSource]] that will be completed when all the stream elements are published.
+    */
+  def toPublisher[F[_], A](
+      stream: Stream[F, A]
+  )(implicit
+      F: Async[F]
+  ): Resource[F, (Publisher[A], DeferredSource[F, Unit])] =
+    StreamUnicastPublisher(stream)
```

Comment:
I actually believe it's unnecessary to expose the `DeferredSource` here. To get access to this information a user can always wrap their `Subscriber` in a `Subscriber` that does something extra for `onComplete` / `onExtra`.

Reply:
While that is technically true, IMHO that would force extra burden on our users. Especially since those functions are not effectual.

Additionally, that approach would make it harder to wait for termination.

However, maybe we should use `Either[Throwable, Unit]` for the type of the `Deferred`?

Reply:
> While that is technically true, IMHO that would force extra burden on our users.

How often will users really need this API? It's an advanced API IMO with rare use-cases. I see how it's useful for the `subscribeStream(...)`, in which case we can use the technique I described in my previous comment.

Reply:
> It's an advanced API IMO

Right, just like the alternative version of `fromPublisher`, where an user could just create its own `Publisher` and manually call `unsafeRunSync()`.
Yet, you were not really happy about that yesterday :)

I agree this is an advanced API and most users should use `subscribe` instead.
But, I would rather try to make this as ergonomic and safe as possible; e.g. actually canceling subsequent `subscribe` calls rather than leaving them in an undefined state as the original implementation did.

Reply:
To clarify: I think `Resource[F, (Publisher[A], DeferredSource[F, Unit])]` where you actually need to use that `Deferred`) is an advanced API. And the API already exposes that information without providing a `DeferredSource`.

Furthermore, if you don't need to use it (i.e. most/all usescases) then it's just plain unergonomic.  `Resource[F, Publisher[A]]` is simple enough.

Reply:
> And the API already exposes that information without providing a DeferredSource.

Sure, just like `fromPublisher(publisher: Publisher[A]): Stream[A]` exposes all that is needed.
So, I don't understand how this is different than adding the `fromPublisher(publisher: Subscriber[A] => F[Unit]): Stream[A]`?

> Furthermore, if you don't need to use it (i.e. most/all usescases)

Then you can just use `streamSubscribe` and simpler than a `Resource` that if you leave without ensuring all the elements were consumed will cancel the downstream consumer.

Reply:
> So, I don't understand how this is different

1. Because there is a performance cost (i.e., it requires `unsafeRunSync()`). In general, it's easier to recover an async effect from an unsafe API than to run an async effect from an unsafe synchronous API.
2. Because there's an existing external usecase in the http4s-jdk-http-client.
3. Because it did not require changing internal implementations, as adding this did.

> Then you can just use `streamSubscribe` and simpler than a `Resource` that if you leave without ensuring all the elements were consumed will cancel the downstream consumer.

What if you actually need a `Publisher`?